sdks/go/pkg/beam/io/textio/textio.go

```diff
@@ -135,55 +134,54 @@ func Write(s beam.Scope, filename string, col beam.PCollection) {
 	s = s.Scope("textio.Write")
 
 	validateScheme(filename)
-	beam.ParDo0(s, &writeFileFn{Filename: filename}, col)
+
+	// NOTE(BEAM-3579): We may never call Teardown for non-local runners and
+	// FinishBundle doesn't have the right granularity. We therefore
+	// perform a GBK with a fixed key to get all values in a single invocation.
+
+	pre := beam.ParDo(s, addFixedKey, col)
+	post := beam.GroupByKey(s, pre)
+	beam.ParDo0(s, &writeFileFn{Filename: filename}, post)
+}
+
+func addFixedKey(elm beam.T) (int, beam.T) {
+	return 0, elm
 }
 
 type writeFileFn struct {
 	Filename string `json:"filename"`
-
-	fs     FileSystem
-	fd     io.WriteCloser
-	writer *bufio.Writer
 }
 
-func (w *writeFileFn) Setup(ctx context.Context) error {
+func (w *writeFileFn) ProcessElement(ctx context.Context, _ int, lines func(*string) bool) error {
 	fs, err := newFileSystem(ctx, w.Filename)
 	if err != nil {
 		return err
 	}
+	defer fs.Close()
+
 	fd, err := fs.OpenWrite(ctx, w.Filename)
 	if err != nil {
 		fs.Close()
 		return err
 	}
+	buf := bufio.NewWriterSize(fd, 1<<20)
```

Comment:
<!--new_thread; commit:21bc8cf0b14f1e76644def3c829bdbe4bda0e675; resolved:0-->
Consider adding a comment about the buffer size. I believe this is 1MB?

Reply:
Done